segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -173,75 +200,68 @@ public int getId() {
      * @param operations The initial set of operations to process (in order). Multiple operations may be processed eventually
      *                   depending on how the operationQueue changes while this is processing.
      */
+    @SneakyThrows
     private void processOperations(Queue<CompletableOperation> operations) {
         log.debug("{}: processOperations (OperationCount = {}).", this.traceObjectId, operations.size());
 
-        // Create a new State and Builder (we need this either initially or after recovery from an error).
-        final QueueProcessingState state = new QueueProcessingState(this.metadataUpdater, this.stateUpdater, this.checkpointPolicy, this.traceObjectId);
-        final DataFrameBuilder<Operation> dataFrameBuilder = new DataFrameBuilder<>(this.durableDataLog, state::commit, state::fail);
+        // Process the operations in the queue. This loop will ensure we do continuous processing in case new items
+        // arrived while we were busy handling the current items.
+        while (!operations.isEmpty()) {
+            try {
+                // Get the DataFrameBuilder, and recover it if necessary.
+                val builder = getDataFrameBuilder(true);
 
-        try {
-            // Process the operations in the queue. This loop will ensure we continue processing after a recoverable failure,
-            // as well as after we processed the entire collection, but found more items in need of processing.
-            while (!operations.isEmpty()) {
                 // Process the current set of operations.
-                processOperations(operations, state, dataFrameBuilder);
+                while (!operations.isEmpty()) {
+                    CompletableOperation o = operations.poll();
+                    if (processOperation(o, builder)) {
+                        // Add the operation as 'pending', only if we were able to successfully append it to a data frame.
+                        // We only commit data frames when we attempt to start a new record (if it's full) or if we try to
+                        // close it, so we will not miss out on it.
+                        this.state.addPending(o);
+                    }
+                }
 
                 // Check if there are more operations to process. If so, it's more efficient to process them now (no thread
                 // context switching, better DataFrame occupancy optimization) rather than by going back to run().
                 if (operations.isEmpty()) {
```

Comment:
There are two while loops. The goal of this is to see if any new operations arrived while we were busy processing the current operations. 

The Outer while loop will execute as long as we keep having new operations in. This is more efficient as it tries to fill up existing DataFrames, it does not deal with thread context switches and it adds virtually no latency.

The inner while loop simply processes all of the current operations. The if block checks to see if we have more operations to fetch (`operationQueue.poll`). If we don't, we flush the current frame out and exit. Otherwise we log a message and go back to the beginning of the outer loop.